freshmaker/pyxis.py

```diff
@@ -0,0 +1,252 @@
+
+import requests
+import logging
+import urllib
+import copy
+from http import HTTPStatus
+from requests_kerberos import HTTPKerberosAuth, OPTIONAL
+from packaging import version
+
+from freshmaker import log, conf
+
+# decrease level of logging for kerberos authentication
+logging.getLogger('requests_kerberos').setLevel('INFO')
+
+
+class PyxisRequestError(Exception):
+    """
+    Error return as a response from Pyxis
+    """
+
+    def __init__(self, status_code, error_response):
+        """
+        Initialize Pyxis request error
+
+        :param int status_code: response status code
+        :param str or dict error_response: response content returned from Pyxis
+        """
+
+        self._status_code = status_code
+        self._raw = error_response
+
+    @property
+    def raw(self):
+        return self._raw
+
+    @property
+    def status_code(self):
+        return self._status_code
+
+
+class Pyxis(object):
+    """ Interface for querying Pyxis"""
+
+    def __init__(self, server_url):
+        self._server_url = server_url
+        # add api version to root url
+        self._api_root = urllib.parse.urljoin(self._server_url, "v1/")
+
+    def _make_request(self, entity, params):
+        """
+        Send a request to Pyxis
+
+        :param str entity: entity part to construct a full URL for request.
+        :param dict params: Pyxis query parameters.
+        :return: Json response from Pyxis
+        :rtype: dict
+        :raises PyxisRequestError: If Pyxis returns error as a response
+        """
+        entity_url = urllib.parse.urljoin(self._api_root, entity)
+
+        auth_method = HTTPKerberosAuth(mutual_authentication=OPTIONAL)
+        response = requests.get(entity_url, params=params, auth=auth_method)
```

Comment:
Please make sure you set a timeout. If you don't, the request will hang forever if the server is unresponsive.